[SYCL] native image handle support for LevelZero. (#8603)
intel/llvm: sycl/doc/extensions/supported/sycl_ext_oneapi_backend_level_zero.md

```
</td>
</tr>
<tr>
<td>image</td>
<td>

``` C++
ze_image_handle_t
```
</td>
<td>

``` C++
struct {
    ze_image_handle_t ZeImageHandle;
    sycl::image_channel_order ChanOrder;
    sycl::image_channel_type ChanType;

>     range<Dimensions> Range;

Reviewer: ```suggestion
    sycl::range<Dimensions> Range;
```
And how is `Dimensions` defined?

Author: The discussion of `make_image` ( starting at line 460 ) has a more full description, explaining why this struct has the members that it has, and what is expected. It felt like that was a more appropriate place for the discussion and I didn't want to duplicate it.  I notice the other interop handles document themselves similarly - with spare descriptions in this table, and more explication lower.

Reviewer: at least adding `sycl::` will not harm, I'll let @gmlueck suggest how we better describe the templated input

Author: Yes, I added the sycl:: to range.

Reviewer 2: > at least adding sycl:: will not harm, I'll let @gmlueck suggest how we better describe the templated input

@smaslov-intel: Can you elaborate on what should be better described?

Reviewer: @gmlueck : the "Dimensions" template argument:

> sycl::range<**Dimensions**> Range;

Reviewer 2: The description of `make_image` says this:

> where the Range should be ordered (width), (width, height), or (width, height, depth) for 1D, 2D and 3D images respectively, with those values matching the dimensions used in the `ze_image_desc` that was used to create the `ze_image_handle_t` initially

Doesn't that adequately describe the `Dimensions`?

Construct a SYCL buffer instance from a pointer to a Level Zero memory allocation. Please refer to <code>make_buffer</code>
description above for semantics and restrictions.
The additional <code>AvailableEvent</code> argument must be a valid SYCL event. The instance of the SYCL buffer class template being constructed must wait for the SYCL event parameter to signal that the memory native handle is ready to be used.
</tr>

<tr>
<td>

``` C++
make_image(
    const backend_input_t<backend::ext_oneapi_level_zero,
                          image<Dimensions, AllocatorT>> &,

>     const context &Context, event AvailableEvent = {})

Reviewer 2: I don't think we should have a default `event` parameter like this.  It seems better to follow `make_buffer` and have two overloads.

Author: Why is that better?   
The `make_buffer` implementation is not so simple. It uses an optional event for `AvailableEvent` in the generic backend declaration, and only the LevelZero specialization has two overloads, and then second overload just calls a helper function with a default event in the "eventless" variant.  
The `make_image` implementation doesn't need a helper, and for that it seems more consistent and simpler to have the optional `AvailableEvent` in all declarations and specializations.

```
</td>
<td>This API is available starting with revision 4 of this specification.

Construct a SYCL image instance from a ze_image_handle_t. 

Because LevelZero has no way of getting image information from an image, it must be provided.  The  <code>backend_input_t</code> is a struct type like so:
``` C++
struct type {
    ze_image_handle_t ZeImageHandle;
    sycl::image_channel_order ChanOrder;
    sycl::image_channel_type ChanType;

>     sycl::range<Dimensions> Range;

Reviewer 2: The SYCL specification [section 4.5.1.3](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:backend-interoperability-make) defines the prototypes for `make_buffer`, and there are two overloads for that API (one with `event` and one without).  That section also specifies the prototypes for `maked_unsampled_image` and `make_sampled_image`, which also have two overloads each.

It is true that `make_image` is not in the SYCL 2020 spec because `image` is a 1.2.1 concept.  However, I think it makes sense for `make_image` to follow the pattern we have established in the spec.

It's also more efficient to have two overloads.  With a single overload, the application needs to construct an `event` whenever it calls `make_image`, even if it doesn't pass one.  I presume this is not a large amount of overhead, but it seems wasteful to construct an `event` object if it isn't needed.

    ext::oneapi::level_zero::ownership Ownership{
        ext::oneapi::level_zero::ownership::transfer};
  };
```

Example Usage
``` C++
sycl::backend_input_t<BE, sycl::image<2>> ImageInteropInput{ ZeHImage, ChanOrder, ChanType, ImgRange_2D, sycl::ext::oneapi::level_zero::ownership::transfer };      
    
auto Image_2D  = sycl::make_image<BE, 2>(ImageInteropInput, Context);
```

>  The input SYCL context <code>Context</code> must be associated with a single device, matching the device used at the prior allocation.

Reviewer 2: ```suggestion
 The input SYCL context <code>Context</code> must be associated with a single device, matching the device used to create the Level Zero image handle.

```

The <code>Context</code> argument must be a valid SYCL context encapsulating a Level-Zero context, and the Level-Zero image must have been created on the same context.
The <code>Ownership</code> input structure member specifies if the SYCL runtime should take ownership of the passed native handle. The default behavior is to transfer the ownership to the SYCL runtime. See section 4.4 for details. If the behavior is "transfer" then the runtime is going to free the input Level-Zero memory allocation. 

> Synchronization rules with this API are described in Section 4.5</td>

Reviewer 2: The description of `make_buffer` contains this sentence:

> Created SYCL buffer can be accessed in another contexts, not only in the provided input context.

Does that apply to `make_image` too?

Author: No.  There was a "multiple device/context" support that was added to the buffers, but I did not add a parallel one for the images as I wanted this first implementation to be simpler. Especially for an interim solution while we wait for SYCL Next.

Reviewer 2: In that case, I think you need to specify this limitation.  For example, add a sentence like:

> The created SYCL image can only be accessed from kernels that are submitted to a queue using this same context.

description above for semantics and restrictions.
The additional <code>AvailableEvent</code> argument must be a valid SYCL event. The instance of the SYCL buffer class template being constructed must wait for the SYCL event parameter to signal that the memory native handle is ready to be used.
</tr>

<tr>
<td>

``` C++
make_image(
    const backend_input_t<backend::ext_oneapi_level_zero,
                          image<Dimensions, AllocatorT>> &,
    const context &Context, event AvailableEvent = {})
```
</td>
<td>This API is available starting with revision 4 of this specification.

Construct a SYCL image instance from a ze_image_handle_t. 

Because LevelZero has no way of getting image information from an image, it must be provided.  The  <code>backend_input_t</code> is a struct type like so:
``` C++
struct type {
    ze_image_handle_t ZeImageHandle;
    sycl::image_channel_order ChanOrder;
    sycl::image_channel_type ChanType;
    sycl::range<Dimensions> Range;
    ext::oneapi::level_zero::ownership Ownership{
        ext::oneapi::level_zero::ownership::transfer};
  };
```

> where the Range should be ordered (width), (width, height), or (width, height, depth) for 1D, 2D and 3D images respectively, 

Reviewer 2: This order seems backwards w.r.t. to other uses of multi-dimensional ranges.  In SYCL , the rightmost dimension varies fastest (see [section 3.11.1](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:multi-dim-linearization)).  Wouldn't this correspond to "width"?

Author: That is not the order images use in SYCL as they were implemented in SYCL 1.2.1  In years past there was a lot of discussion about this, because it was weird for the ranges to be interpreted one way for buffers but another for images. That's just the way it ended up. IIRC, the computcpp implementation did the same thing. Haven't checked it recently.

I will add a note that this interpretation of range may be interim.

Since we are talking about it, I find the terminology "dimension that varies fastest" to be confusing. Wouldn't that be interpreted as the z dimension, not the x?  Because a +1 change in the z would cause the biggest jump in the linear position.  Talking about ranges abstractly is always challenging in the context of buffers and images, where they are no longer abstract.  I prefer the terminology of "contiguous memory". Enumerating the x term results in visiting contiguous memory. And, even better, I think we shouldn't shy away from using depth/z, height/y, width/x when talking about buffers and images.  That's the metaphor users are using in their head, the docs should follow suit.   Besides, our ranges aren't truly N-space anyway, are they? There is no `range<4>`  .

Reviewer 2: Yes, I agree.  It's more important that the `range` order matches the order we have already in the `image` constructor.  If the `image` constructor uses the the first dimension for "width" we should do the same here.

with those values matching the dimensions used in the `ze_image_desc` that was used to create the `ze_image_handle_t` initially. 

Example Usage
``` C++

> sycl::backend_input_t<BE, sycl::image<2>> ImageInteropInput{ ZeHImage, ChanOrder, ChanType, ImgRange_2D, sycl::ext::oneapi::level_zero::ownership::transfer };      

Reviewer 2: ```suggestion
sycl::backend_input_t<BE, sycl::image<2>> ImageInteropInput{ ZeHImage, ChanOrder, ChanType, ImgRange_2D };      
```

Is this better?  Since `transfer` is the default, it doesn't need to be specified.  Or maybe you wanted to show it intentionally for emphasis?

Author: I don't have an opinion here.  However, that said, the ownership value is super important. If you get it wrong, you'll get crashes and/or leaks.  I used a default value because that's what we do in `make_buffer`, but if it weren't for that I don't think I'd have a default - I'd make the user specify it each time.

> auto Image_2D  = sycl::make_image<BE, 2>(ImageInteropInput, Context);

Reviewer 2: ```suggestion
auto Image_2D  = sycl::make_image<BE, sycl::image<2>>(ImageInteropInput, Context);
```

I think, right?

Author: No.  It is correct as written.  I could lose the `auto` though.  That might make it clearer.
`sycl::image<2> Image_2D = sycl::make_image<BE, 2>(ImageInteropInput, Context);`

Reviewer 2: I see.  I think it would help if you changed the prototype of `make_image` in the left column of this table to show the template parameters.  For example:

```
template<backend Backend, int Dimensions = 1, typename AllocatorT = sycl::image_allocator>
image<Dimensions, AllocatorT> make_image(
    const backend_input_t<backend::ext_oneapi_level_zero,
                          image<Dimensions, AllocatorT>> &,
    const context &Context, event AvailableEvent)
```

I'm assuming you have a template parameter `AllocatorT` like this?  I also assume the `Dimensions` template parameter defaults to `1` to match the SYCL 1.2.1 spec, where the `image` object also defaults to 1 dimension?

```

 The input SYCL context <code>Context</code> must be associated with a single device, matching the device used to create the Level Zero image handle.
The <code>Context</code> argument must be a valid SYCL context encapsulating a Level-Zero context, and the Level-Zero image must have been created on the same context.
The <code>Ownership</code> input structure member specifies if the SYCL runtime should take ownership of the passed native handle. The default behavior is to transfer the ownership to the SYCL runtime. See section 4.4 for details. If the behavior is "transfer" then the runtime is going to free the input Level-Zero memory allocation. 

> Synchronization rules with this API are described in Section 4.5</td>

Reviewer 2: Sorry, I just noticed this last sentence:

> Synchronization rules with this API are described in Section 4.5

That section is quite specific to buffers, talking about the effect on the underlying USM, etc.

What do we need to say about the synchronization rules for interop images?  I think we need to say that the image destructor blocks only when ownership is `keep`.  Is there anything more we need to say?  If that's it, we might not need a separate section, and we can just have a new paragraph at the end of the `make_image` description.

description above for semantics and restrictions.
The additional <code>AvailableEvent</code> argument must be a valid SYCL event. The instance of the SYCL buffer class template being constructed must wait for the SYCL event parameter to signal that the memory native handle is ready to be used.
</tr>

<tr>
<td>

``` C++

> make_image(

Reviewer 2: I think you should show the template parameters for the `make_image` functions.  I know we don't shows them for the others, but it's less critical for those functions because the template parameters are specified in the SYCL 2020 spec.  However, SYCL 2020 doesn't have `make_image`, so there is currently no place where those template parameters are specified.

If you wanted to be consistent, we could add the template parameters for all these `make_OBJECT` functions.

> make_image<backend::ext_oneapi_level_zero, Dims>(

Reviewer 2: This does not address my previous comment about showing the template parameters.  Because `make_image` is not in the SYCL 2020 specification, there is no specification of the template parameters or their default values.  Even the return type of this function is not specified.

An additional problem is that all of the entries in this table are a sort of pseudo code.  They aren't template definitions because they don't show the `tempalate` keyword and template arguments.  They are sort of a pseudo-code version of a template specialization, but the syntax isn't correct.

I think the best solution is to change all the entries in the first column to be real template definitions.  Your entries for `make_image` would look like:

```
template<backend Backend, int Dimensions = 1, typename AllocatorT = sycl::image_allocator>
image<Dimensions, AllocatorT> make_image(
    const backend_input_t<Backend, image<Dimensions, AllocatorT>> &backendObject,
    const context &targetContext);

template<backend Backend, int Dimensions = 1, typename AllocatorT = sycl::image_allocator>
image<Dimensions, AllocatorT> make_image(
    const backend_input_t<Backend, image<Dimensions, AllocatorT>> &backendObject,
    const context &targetContext, event availableEvent);
```

(I'm guessing at the default template parameters here, so correct them if necessary.)

To make the entries in the table consistent, you should also change the other entries to be real template definitions.  The synopses can just be copied from the SYCL 2020 specification.

Author: done

  };
```
where the Range should be ordered (width), (width, height), or (width, height, depth) for 1D, 2D and 3D images respectively, 
with those values matching the dimensions used in the `ze_image_desc` that was used to create the `ze_image_handle_t` initially. 
Note that the range term ordering (width first, depth last) is true for SYCL 1.2.1 images that are supported here. But future classes like
sampled_image and unsampled_image might have a different ordering.

Example Usage
``` C++
sycl::backend_input_t<BE, sycl::image<2>> ImageInteropInput{ ZeHImage, ChanOrder, ChanType, ImgRange_2D, sycl::ext::oneapi::level_zero::ownership::transfer };      
    
sycl::image<2> Image_2D  = sycl::make_image<BE, 2>(ImageInteropInput, Context);
```

 The input SYCL context <code>Context</code> must be associated with a single device, matching the device used to create the Level Zero image handle.
The <code>Context</code> argument must be a valid SYCL context encapsulating a Level-Zero context, and the Level-Zero image must have been created on the same context. The created SYCL image can only be accessed from kernels that are submitted to a queue using this same context.

> The <code>Ownership</code> input structure member specifies if the SYCL runtime should take ownership of the passed native handle. The default behavior is to transfer the ownership to the SYCL runtime. See section 4.4 for details. If the behavior is "transfer" then the SYCL runtime is going to free the input Level-Zero memory allocation, meaning the memory will be freed when the ~image destructor fires.  If the behavior is "keep", then the memory will not be freed by the ~image destructor, and it is the responsibility of the caller to free the memory appropriately. 

Reviewer 2: Repeating an earlier comment here so it does not get lost ... there should be some description of the synchronization rules here.  Does the image destructor block when it is created with this interop API?  In the buffer case, the buffer destructor blocks only if the ownership is "keep".  It probably makes sense to have that semantic here also.

description above for semantics and restrictions.
The additional <code>AvailableEvent</code> argument must be a valid SYCL event. The instance of the SYCL buffer class template being constructed must wait for the SYCL event parameter to signal that the memory native handle is ready to be used.
</tr>

<tr>
<td>

``` C++
template<backend Backend, int Dimensions = 1, typename AllocatorT = sycl::image_allocator>
image<Dimensions, AllocatorT> make_image(
    const backend_input_t<Backend, image<Dimensions, AllocatorT>> &backendObject,
    const context &targetContext);

> ```

Reviewer 2: Thanks for adding the template definition.  This makes the prototype of the function much clearer.  Two comments, though:

1. The HTML render on my monitor (1920 x 1200) has a horizontal scroll bar, which makes the table hard to read.  (You need to scroll left and right to read the "description" column.)  This is caused because the synopses for `make_image` use more columns than the other synopses in this table.  Can you add some line breaks to these synopses to avoid this?  For example, maybe:

   ```
   template<backend Backend, int Dimensions = 1,
            typename AllocatorT = sycl::image_allocator>
   image<Dimensions, AllocatorT> make_image(
       const backend_input_t<Backend,
                             image<Dimensions, AllocatorT>> &backendObject,
       const context &targetContext);
   ```

2. Please update the other entries in the table to have the `template` keyword, template parameters, and return type to match the style of the `make_image` entries.

Author: FWIW, my HTML preview in VS code does not insert scroll bars. I'll add more line breaks though.

</td>
<td>This API is available starting with revision 5 of this specification.

Construct a SYCL image instance from a ze_image_handle_t. 

Because LevelZero has no way of getting image information from an image, it must be provided.  The  <code>backend_input_t</code> is a struct type like so:
``` C++
struct type {
    ze_image_handle_t ZeImageHandle;
    sycl::image_channel_order ChanOrder;
    sycl::image_channel_type ChanType;
    sycl::range<Dimensions> Range;
    ext::oneapi::level_zero::ownership Ownership{
        ext::oneapi::level_zero::ownership::transfer};
  };
```
where the Range should be ordered (width), (width, height), or (width, height, depth) for 1D, 2D and 3D images respectively, 
with those values matching the dimensions used in the `ze_image_desc` that was used to create the `ze_image_handle_t` initially. 
Note that the range term ordering (width first, depth last) is true for SYCL 1.2.1 images that are supported here. But future classes like
sampled_image and unsampled_image might have a different ordering.

Example Usage
``` C++

> sycl::backend_input_t<BE, sycl::image<2>> ImageInteropInput{ 

Author: done

    ZeHImage, 
    ChanOrder,
    ChanType, 

>     ImgRange_2D, 

Author: i added all

    sycl::ext::oneapi::level_zero::ownership::transfer };      
    
sycl::image<2> Image_2D  
  = sycl::make_image<BE, 2>(ImageInteropInput, Context);
```

>  The input SYCL context <code>Context</code> must be associated with a single device, matching the device used to create the Level Zero image handle.

Reviewer: > must be associated with a single device, matching the device used to create the Level Zero image handle.

NIT: this is an overkill to require that the context has that single device only. It is enough that the image is only used on the device where it was created, but not on other devices on the same context (limitation that exists today, but can be relaxed in future)

The <code>Ownership</code> input structure member specifies if the SYCL runtime should take ownership of the passed native handle. The default behavior is to transfer the ownership to the SYCL runtime. See section 4.4 for details. If the behavior is "transfer" then the SYCL runtime is going to free the input Level-Zero memory allocation, meaning the memory will be freed when the ~image destructor fires. When using "transfer" the ~image destructor may not need to block.  If the behavior is "keep", then the memory will not be freed by the ~image destructor, and the ~image destructor blocks until all work in the queues on the image have been completed. When using "keep" it is the responsibility of the caller to free the memory appropriately. 
</td>
</tr>

<tr>
<td>

``` C++
template<backend Backend, int Dimensions = 1, typename AllocatorT = sycl::image_allocator>
image<Dimensions, AllocatorT> make_image(
    const backend_input_t<Backend, image<Dimensions, AllocatorT>> &backendObject,
    const context &targetContext, event availableEvent);
```
</td>

> <td>This API is available starting with revision 4 of this specification.

Reviewer 2: ```suggestion
<td>This API is available starting with revision 5 of this specification.

```
